pil-analyzer/src/type_inference.rs

```diff
@@ -638,17 +643,70 @@ impl<'a> TypeChecker<'a> {
                             self.expect_type_of_pattern(&value_type, pattern)?;
                         }
                         StatementInsideBlock::Expression(expr) => {
-                            self.expect_type_with_flexibility(self.statement_type, expr)?;
+                            self.check_kind_in_expression(expr)?
                         }
                     }
                 }
                 let result = self.infer_type_of_expression(expr);
+                let result = self.check_kind(result?);
                 self.local_var_types.truncate(original_var_count);
                 result?
             }
         })
     }
 
+    fn check_kind_in_expression(
+        &mut self,
+        expr: &mut powdr_ast::parsed::Expression<Reference>,
+    ) -> Result<(), String> {
+        let empty_tuple = Type::Tuple(TupleType { items: vec![] });
+        match self.lambda_kind {
+            FunctionKind::Constr => {
+                if self.expect_type(&self.statement_type.ty, expr).is_ok() {
```

Comment:
If possible, call `infer_type` and then check if the type equals `()`, otherwise try the other type